### ISSUE TITLE
[ADAP-1077] [Drops Table In Separate Transaction]  - Tables are dropped then the _tmp table renamed, causing other queries referencing the table to fail

### ISSUE THREAD

**Issue by gtmehrer** [dbt-labs/dbt-redshift#693]
### Is this a new bug in dbt-redshift?

- [X] I believe this is a new bug in dbt-redshift
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Models with table materialization are dropped then the transaction is committed.  This causes queries referencing the refreshing model to fail if they run at the same time since the table doesn't exist.

I've tried setting `autocommit: true ` in profiles.yml to no effect.

### Expected Behavior

The old behavior was to drop/rename in one transaction.

### Steps To Reproduce

Run `dbt run` command in dbt-redshift >=1.7.1 (core >=1.7.3) against a table model.  

dbt-redshift 1.6.5 (core 1.6.9)  runs in a single transaction

### Relevant log output

```shell
[0m08:54:14.597104 [debug] [Thread-2  ]: On model.$$$$: BEGIN
[0m08:54:14.646101 [debug] [Thread-2  ]: SQL status: SUCCESS in 0.0 seconds
[0m08:54:14.669096 [debug] [Thread-2  ]: Applying DROP to: $$$$
[0m08:54:14.671270 [debug] [Thread-2  ]: Using redshift connection "model.$$$$"
[0m08:54:14.672171 [debug] [Thread-2  ]: On model.$$$$: /* {"app": "dbt", "dbt_version": "1.7.3", "profile_name": "redshift", "target_name": "$$$$", "node_id": "model.$$$$"} */
drop table if exists $$$$ cascade
[0m08:54:14.719695 [debug] [Thread-2  ]: SQL status: SUCCESS in 0.0 seconds
[0m08:54:14.723391 [debug] [Thread-2  ]: On model.$$$$: COMMIT
```

### Environment

```markdown
- OS: macOS sonoma
- Python: 3.8.13
- dbt-core: 1.7.4
- dbt-redshift: 1.7.1
```

**Comment by mikealfare** [comment]
This is a regression that requires some design, hence the `refinement` tag. I believe the issue is that we moved from `psycopg2` to `redshift-connector`. `redshift-connector` implements DBAPI2.0 correctly, which assumes execution of statements one at a time (you need to build the transaction manually if you want it). Whereas `psycopg2` allows you to pass it multiple statements and it will automatically "wrap" it in a transaction. The assumption of the latter scenario exists in many places throughout `dbt-redshift`. This change was made in `1.5`, so I'm a little surprised that this was working as a single transaction in `1.6`. Perhaps it's the interaction of this scenario, along with the implementation of the `replace` set of macros that handles swapping tables with materialized views and vice versa.

**Comment by martynydbt** [comment]
@VersusFacit i added a backport label!

**Comment by VersusFacit** [comment]
@gtmehrer Hey there! Thanks for the submission. Just an upfront disclaimer that autocommit is `True` by default.

I recreated your logs essentially as you described.

My model:
```
{{
config(
    materialized='table'
)
}}

select 2
```

### The Logs

dbt-redshift 1.6.6
```
drop table if exists ***."dbt_mila"."basic__dbt_backup" cascade
```

dbt-redshift 1.7.3 
```
drop table if exists ***."dbt_mila"."basic" cascade
```

But with `autocommit: false`, it seems you have the option for your particular workflow to obtain what you need.

dbt-redshift 1.6.6
```
drop table if exists "ci"."dbt_mila"."basic__dbt_backup" cascade
```

dbt-redshift 1.7.3
```
drop table if exists "ci"."dbt_mila"."basic__dbt_backup" cascade
```

### Request for more info

Could you try with `autocommit: False` to see if that addresses your issue?

> This causes queries referencing the refreshing model to fail if they run at the same time since the table doesn't exist.

What is your use case by chance? It's not immediately clear what your error is from this description or how it's being triggered, but I'd love to know more.

As in you have queries going that are looking for the model? If your workflow for this model uses `autocommit: False` even just as a one-off, that might just work!

**Comment by VersusFacit** [comment]
There is an alternate possibility the team is exploring: It could be we need to remove some BEGIN/COMMIT barriers in special edge cases like this one you may have found. I have a PR up for that just in case. We'd still very much appreciate you trying the autocommit adjustment as a sanity check and providing more insight into your use case 🖖

**Comment by mikealfare** [comment]
@VersusFacit and I talked through this quite a bit and we're pretty sure we found the issue. The root cause is that we are setting a class variable on `BaseRelation` in such a way that it does not get properly updated at the implementation layer.

## Diagnosis

We set `renameable_relations` in `dbt-core` here:
https://github.com/dbt-labs/dbt-core/blob/0a6d0c158e5a1956bd48e53793ca3b28faee2b34/core/dbt/adapters/base/relation.py#L44

and attempt to override it with the correct configuration for `dbt-redshift` here:
https://github.com/dbt-labs/dbt-redshift/blob/d77f5ee9561bf11eae56b2832340e3c11005b6f9/dbt/adapters/redshift/relation.py#L33

Unfortunately, because `BaseRelation` is a frozen dataclass, the override does not take affect. We didn't notice this in `1.6` because `dbt-redshift` uses the default table materialization in `1.6`. However, we needed to use the renameable relations concept in `1.7` because of interactions between tables and materialized views. This required us to override the default table materialization to update this logic:
https://github.com/dbt-labs/dbt-core/blob/0a6d0c158e5a1956bd48e53793ca3b28faee2b34/core/dbt/include/global_project/macros/materializations/models/table.sql#L40

to include this conditional:
https://github.com/dbt-labs/dbt-redshift/blob/d77f5ee9561bf11eae56b2832340e3c11005b6f9/dbt/include/redshift/macros/materializations/table.sql#L40-L44

The issue is that `existing_relation.renameable_relations` is always empty, hence `existing_relation.can_be_renamed` is always `False`. The correct logic path passes the conditional, which would result in the original behavior from `dbt-core`. Instead, we run through the `else` path and drop the relation.

## Demonstrating Test

The following test in `dbt-redshift` demonstrates the issue:
```python
from dbt.adapters.redshift.relation import RedshiftRelation
from dbt.contracts.relation import RelationType


def test_renameable_relation():
    relation = RedshiftRelation.create(
        database="my_db",
        schema="my_schema",
        identifier="my_table",
        type=RelationType.Table,
    )
    assert relation.renameable_relations == frozenset({
        RelationType.View,
        RelationType.Table
    })
```
This test currently fails because `relation.renameable_relations==()`.

## Fix

We need to update this line:
https://github.com/dbt-labs/dbt-core/blob/0a6d0c158e5a1956bd48e53793ca3b28faee2b34/core/dbt/adapters/base/relation.py#L44
to use the `field` method like this:
https://github.com/dbt-labs/dbt-core/blob/0a6d0c158e5a1956bd48e53793ca3b28faee2b34/core/dbt/adapters/base/relation.py#L36
Since the default is a built-in (`tuple`), you should probably be able to avoid the lambda and do:
```python
renameable_relations: SerializableIterable = field(default_factory=tuple)
```
While we're here, the same mistake was made in a related area on the next line of code:
https://github.com/dbt-labs/dbt-core/blob/0a6d0c158e5a1956bd48e53793ca3b28faee2b34/core/dbt/adapters/base/relation.py#L50
This should also use the `field` method to set the empty tuple as a default. The test should be updated (or a second one should be created) to check this is configured correctly as well.

The fix needs to be implemented in `dbt-core>=1.6` and `dbt-adapters>=1.8`. The tests need to be implemented in `dbt-snowflake>=1.6`, `dbt-redshift>=1.6`, and `dbt-bigquery>=1.7`.

## Risks

We've been operating under the assumption that we've been using the `renameable_relations` and `replaceable_relations` as expected. However, since these were both always `False`, we may be introducing new logic paths by fixing this bug. Put another way, resolving this bug is effectively implementing `renameable_relations`, `replaceable_relations`, and the associated `replace` macro in some places.

**Comment by dbeatty10** [comment]
That is some great info @mikealfare and @VersusFacit 🧠

While we're in this area of code, here's one other thing that @VersusFacit and I noticed in https://github.com/dbt-labs/dbt-redshift/issues/693#issuecomment-1964399485:
- in the last statement, `my_model__dbt_backup` is dropped if it exists.

It doesn't appear to ever get created, so it's a no-op. But it looks like it may be an unnecessary statement.

Maybe it will be naturally handled along with the other fixes?

**Comment by VersusFacit** [comment]
I dont think it will be since it's an always drop. I think it's just a leftover artifact. I can make a trick to track dealing with this at some point :)

**Comment by gtmehrer** [comment]
Thanks so much for digging into this!  @VersusFacit  I did try with autocommit set to both True and False with the same results either way.  

Our use case is that we refresh many models frequently throughout the day, around every 30 mins.  Outside processes querying dbt managed tables will fail when they attempt to reference a table that has been dropped (when the transaction is immediately commited by dbt).  The table is generally only gone for a few seconds before it's recreated by another transaction so it's a little difficult to replicate without high query volume.  

Also, it just occurred to me that we are using Redshift's snapshot isolation, which may result in a different error than serializable isolation would.  https://docs.aws.amazon.com/redshift/latest/dg/r_STV_DB_ISOLATION_LEVEL.html

Let me know if I can provide more info or help with testing!

**Comment by mikealfare** [comment]
We merged the fix today. We'll need to do a patch release before it's available on PyPI, which should happen in the next week. Feel free to install from branch `1.7.latest` if you want to test it in the meantime.

Closing as resolved.

**Comment by AlessandroLollo** [comment]
@mikealfare has the fix been included in `dbt-redshift==1.7.5`?
Thanks! 🙌

**Comment by mikealfare** [comment]
> @mikealfare has the fix been included in `dbt-redshift==1.7.5`? Thanks! 🙌

It is. We [published](https://pypi.org/project/dbt-redshift/1.7.5/) a few hours ago.

**Comment by AlessandroLollo** [comment]
Thank you @mikealfare 🙌
